File: stackdio/server/core/static/stackdio/api/Formulas.js
/*!
  * Copyright 2014,  Digital Reasoning
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  * 
*/

define(['q', 'settings', 'model/models'], function (Q, settings, models) {
    var self = this;

    return {
        load: function () {
            var deferred = Q.defer();

            $.ajax({
                url: settings.api.formulas.formulas,
                type: 'GET',
                headers: {
                    'Accept': 'application/json'
                },
                success: function (response) {
                    var formulas = response.results.map(function (formula) {
                        return new models.Formula().create(formula);
                    });
                    deferred.resolve(formulas);
                },
                error: function (request, status, error) {
                    deferred.reject(new Error(error));
                }
            });

            return deferred.promise;
        },
        load_global: function () {
            var deferred = Q.defer();

            $.ajax({
                url: settings.api.admin.global_orchestration_formulas,
                type: 'GET',
                headers: {
                    'Accept': 'application/json'
                },
                success: function (response) {
                    var formulas = response.results.map(function (formula) {
                        return new models.Formula().create(formula);
                    });
                    deferred.resolve(formulas);
                },
                error: function (request, status, error) {
                    deferred.reject(new Error(error));
                }
            });

            return deferred.promise;
        },
        import: function (uri, git_username, git_password) {
            var deferred = Q.defer();

            $.ajax({
                url: settings.api.formulas.formulas,
                type: 'POST',
                data: JSON.stringify({
                    uri: uri,
                    git_username: git_username,
                    git_password: git_password
                }),
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': stackdio.settings.csrftoken,
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve(new models.Formula().create(data));
                },
                error: function (request, status, error) {
                    var response = '';

                    try {
                        response = JSON.parse(request.responseText).detail;
                    } catch (ex) {
                        response = request.responseText.substr(0,100);
                    }
                    deferred.reject(response);
                }
            });

            return deferred.promise;
        },
        import_global: function (uri, git_username, git_password) {
            var deferred = Q.defer();

            $.ajax({
                url: settings.api.admin.global_orchestration_formulas,
                type: 'POST',
                data: JSON.stringify({
                    uri: uri,
                    git_username: git_username,
                    git_password: git_password
                }),
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': stackdio.settings.csrftoken,
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve(new models.Formula().create(data));
                },
                error: function (request, status, error) {
                    var response = '';

                    try {
                        response = JSON.parse(request.responseText).detail;
                    } catch (ex) {
                        response = request.responseText.substr(0,100);
                    }
                    deferred.reject(response);
                }
            });

            return deferred.promise;
        },
        delete: function (formula) {
            var deferred = Q.defer();

            $.ajax({
                url: formula.url,
                type: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': stackdio.settings.csrftoken,
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve();
                },
                error: function (request, status, error) {
                    deferred.reject(JSON.parse(request.responseText).detail);
                }
            });

            return deferred.promise;
        },
        getFormula: function (formulaId) {
            var deferred = Q.defer();

            $.ajax({
                url: settings.api.formulas.formulas + formulaId + '/',
                type: 'GET',
                headers: {
                    'Accept': 'application/json'
                },
                success: function (response) {
                    deferred.resolve(new models.Formula().create(response));
                },
                error: function (request, status, error) {
                    deferred.reject(new Error(error));
                }
            });

            return deferred.promise;
        },
        getProperties: function (formula) {
            var deferred = Q.defer();
            var self = this;

            $.ajax({
                url: formula.properties,
                type: 'GET',
                headers: {
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve(data.properties);
                },
                error: function (request, status, error) {
                    deferred.reject(new Error(error));
                }
            });

            return deferred.promise;
        },
        update: function (formula) {
            var deferred = Q.defer();
            var self = this;

            $.ajax({
                url: formula.url,
                type: 'PUT',
                data: JSON.stringify({public: !formula.public}),
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': stackdio.settings.csrftoken,
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve();
                },
                error: function (request, status, error) {
                    deferred.reject(JSON.parse(request.responseText).detail);
                }
            });

            return deferred.promise;
        },
        updateFromRepo: function (formula, git_password) {
            var deferred = Q.defer();

            $.ajax({
                url: formula.action,
                type: 'POST',
                data: JSON.stringify({
                    action: 'update',
                    git_password: git_password
                }),
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': stackdio.settings.csrftoken,
                    'Accept': 'application/json'
                },
                success: function (data, status, response) {
                    deferred.resolve();
                },
                error: function (request, status, error) {
                    deferred.reject(JSON.parse(request.responseText).detail);
                }
            });

            return deferred.promise;
        }
    }
});

File: stackdio/ui/views/formulas.py
# -*- coding: utf-8 -*-

# Copyright 2016,  Digital Reasoning
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from django.http import Http404
from django.shortcuts import get_object_or_404

from stackdio.api.formulas.models import Formula
from stackdio.ui.views import PageView, ModelPermissionsView, ObjectPermissionsView
from stackdio.ui.utils import get_object_list


class FormulaImportView(PageView):
    template_name = 'formulas/formula-import.html'
    viewmodel = 'viewmodels/formula-import'

    def get(self, request, *args, **kwargs):
        if not request.user.has_perm('formulas.create_formula'):
            # No permission granted
            raise Http404()
        return super(FormulaImportView, self).get(request, *args, **kwargs)


class FormulaListView(PageView):
    template_name = 'formulas/formula-list.html'
    viewmodel = 'viewmodels/formula-list'

    def get_context_data(self, **kwargs):
        context = super(FormulaListView, self).get_context_data(**kwargs)
        context['has_admin'] = self.request.user.has_perm('formulas.admin_formula')
        context['has_create'] = self.request.user.has_perm('formulas.create_formula')
        context['object_list'] = get_object_list(self.request.user, Formula)
        return context


class FormulaModelPermissionsView(ModelPermissionsView):
    viewmodel = 'viewmodels/formula-model-permissions'
    model = Formula


class FormulaDetailView(PageView):
    template_name = 'formulas/formula-detail.html'
    viewmodel = 'viewmodels/formula-detail'
    page_id = 'detail'

    def get_context_data(self, **kwargs):
        context = super(FormulaDetailView, self).get_context_data(**kwargs)
        pk = kwargs['pk']
        # Go ahead an raise a 404 here if the formula doesn't exist rather
        # than waiting until later.
        formula = get_object_or_404(Formula.objects.all(), pk=pk)
        if not self.request.user.has_perm('formulas.view_formula', formula):
            raise Http404()
        context['formula'] = formula
        context['has_admin'] = self.request.user.has_perm('formulas.admin_formula', formula)
        context['has_delete'] = self.request.user.has_perm('formulas.delete_formula', formula)
        context['has_update'] = self.request.user.has_perm('formulas.update_formula', formula)
        context['page_id'] = self.page_id
        return context


class FormulaObjectPermissionsView(ObjectPermissionsView):
    template_name = 'formulas/formula-object-permissions.html'
    viewmodel = 'viewmodels/formula-object-permissions'
    page_id = 'permissions'

    def get_context_data(self, **kwargs):
        context = super(FormulaObjectPermissionsView, self).get_context_data(**kwargs)
        pk = kwargs['pk']
        # Go ahead an raise a 404 here if the formula doesn't exist rather
        # than waiting until later.
        formula = get_object_or_404(Formula.objects.all(), pk=pk)
        if not self.request.user.has_perm('formulas.admin_formula', formula):
            raise Http404()
        context['formula'] = formula
        context['has_admin'] = self.request.user.has_perm('formulas.admin_formula', formula)
        context['page_id'] = self.page_id
        return context

    def get_object(self):
        return get_object_or_404(Formula.objects.all(), pk=self.kwargs['pk'])


class FormulaPropertiesView(FormulaDetailView):
    template_name = 'formulas/formula-properties.html'
    viewmodel = 'viewmodels/formula-properties'
    page_id = 'properties'
